Refactor useListSelect hook

JesseSDevaney opened the pull request:
**Problem:** The `useListSelect` hook felt non-intuitive when toggling multiple items, so I wanted to refactor it to make it easier to use and understand.

**Context:**
- I use this hook in the refactoring of https://github.com/metabase/metabase/issues/24018. I created a separate PR to make it easier to review the changes in the hook and its consumers.

---
**Desired Behavior**/**Acceptance Criteria**
- [ ] implement a function to select multiple items (reset selected items to only these items): `selectOnlyTheseItems(items)`
- [ ] refactor consumers of `toggleAll(items)` to use `selectOnlyTheseItems(items)` instead
  - _why:_ `toggleAll(items)` would select "non-selected" items while at the same time unselect "selected " items. This made the consumer responsible of keeping track of what was selected and then remove those items from the list they keep track off in order to select all items. Not very consumer friendly.
- the consumers have no altered behaviors (in this case only `CollectionContent`)
  - [ ] Collection selections behave the same before and after the updates
    ![useListSelect](https://github.com/metabase/metabase/assets/22608765/c9f8ff58-ce6d-48ce-aa30-336431fc14f0)

---
**How to Verify**

1. Open up a collection that has items in it
2. Test out selecting and unselecting items
3. Test out selecting-all and unselecting-all items
4. Verify that it behaves in the way you would expect

JesseSDevaney committed:
refactor use-list-select hook

- toggleAll(items) felt unintuitive as some items could be selected while simultaneously unselecting others
  - switched to selectOnlyTheseItems(items) to be more explicit on what the function is doing
- updated consumers of useListSelect to use the new API

npfitz reviewed (changes requested):
I think I get the purpose of the change. I do wonder if having simple functions to enable and disable everything in the list would be easier (for both the hook and the consumer), but I've actually never interacted with the hook so I'll leave that to your judgement 👍 

Regardless, it would be good to have a few unit tests for the hook though. Do you mind writing some up?

JesseSDevaney committed:
convert to typescript

JesseSDevaney committed:
improve types

JesseSDevaney committed:
add desired behavior verification for useListSelect

JesseSDevaney commented:
@npfitz It is ready for re-review - thank you for suggesting to add tests for the hook :heart: 

:heavy_check_mark: I converted the hook to TypeScript and added tests to verify the hook's desired behavior

npfitz commented:
Awesome! Looks good @JesseSDevaney
